### PR TITLE
fix for invalid texture

### PR DESCRIPTION
**Pull request by GoodBoyDigital**
does what it says on the tin!
Please give this a test!

### CODE REVIEW

**Comment by GoodBoyDigital** [comment]
Totally agree! Just wanted to see if this solves the flickering issue first.

**Comment by finscn** [comment]
In my MacBook , 
the   `gl.getParameter(gl.MAX_TEXTURE_IMAGE_UNITS)`==16 ,  `settings.SPRITE_MAX_TEXTURES`==32 .

So I think the correct way is using these codes : 
```
const glMaxTextures = gl.getParameter(gl.MAX_TEXTURE_IMAGE_UNITS);
const maxTextures = Math.min(glMaxTextures, settings.SPRITE_MAX_TEXTURES);
```

.

**Comment by finscn** [comment]
I test my case , still has 
```
WebGL: INVALID_ENUM: activeTexture: texture unit out of range
```

**Comment by finscn** [comment]
I think the key point is not  the number of `maxTextures`.

It's  about  https://github.com/pixijs/pixi.js/issues/3373#issuecomment-265764333 

* why group.ids[j] could be -1 ? 
* why all `t._enabled === TICK` ?  https://github.com/pixijs/pixi.js/blob/dev/src/core/sprites/webgl/SpriteRenderer.js#L275-L295

**Comment by GoodBoyDigital** [comment]
Interesting! Please can you share the code for your case?

**Comment by finscn** [comment]
@GoodBoyDigital  , my test case is very complex , it's a real game project.

It will render many sprites, meshes, graphics (include rect circle etc.)   in one game tick.

**Comment by finscn** [comment]
after  https://github.com/pixijs/pixi.js/pull/3474/files#diff-b87a2a9840fecac0080b87cdd72de184R93 
you miss a  `break;` , I think

**Comment by GoodBoyDigital** [comment]
Good catch! Will tweak when I get home :)

**Comment by finscn** [comment]
@GoodBoyDigital  , the new patch still didn't fix this bug.

**Comment by finscn** [comment]
For my test case , this error message is throw at WebGLRenderer.prototype.bindTexture
```
            // bind the current texture
            this.boundTextures[location] = texture;

// ****** this line ******
            gl.activeTexture(gl.TEXTURE0 + location);

            gl.bindTexture(gl.TEXTURE_2D, glTexture.texture);
```

**Comment by finscn** [comment]
#3474

**Comment by GoodBoyDigital** [comment]
Thanks for letting me know @finscn ! 
The issue occurs when the same texture is bound to two different texture locations. This situation can result in -1 being passed in as the texture location. Getting closer though! :)

If you could share you code I would be happy to debug. That way I can get to the bottom of the issue much quicker! (maybe on a private channel if its sensitive stuff?)

**Comment by finscn** [comment]
@GoodBoyDigital  , I've modified my reply . Your patch is OK for me.

**Comment by GoodBoyDigital** [comment]
wooo! 👍

**Comment by finscn** [comment]
@GoodBoyDigital  I think it's time to merge some PRs  which have been  approved , and release a new version of PIXI.js  as a gift for  New Year.

BTW : why there is an error of travis in this PR ? Could you fix it ?

And 

Happy New Year.

**Comment by GoodBoyDigital** [comment]
@staff0rd fixed :D

**Comment by bigtimebuddy** [comment]
@ivanpopelyshev do you have any follow-up here?

File: src/core/renderers/webgl/TextureManager.js [hunk 1]
```diff
@@ -79,6 +78,29 @@ export default class TextureManager
             return null;
         }
 
+        const boundTextures = this.renderer.boundTextures;
+
+        // if the location is undefined then this may have been called by n event.
+        // this being the case the texture may already be bound to a slot. As a texture can only be bound once
+        // we need to find its current location if it exists.
+        if (location === undefined)
```
Comment (englercj) [inline src/core/renderers/webgl/TextureManager.js]: Should we do `== null` here to catch `undefined` and `null`?

File: src/core/renderers/webgl/TextureManager.js [hunk 2]
```diff
@@ -79,6 +78,29 @@ export default class TextureManager
...
+        {
+            location = 0;
+
+            // TODO maybe we can use texture bound ids later on...
+            // check if texture is already bound..
+            for (let i = 0; i < boundTextures.length; i++)
```
Comment (englercj) [inline src/core/renderers/webgl/TextureManager.js]: `++i` :)

**Pull request merged**